_posts/2016-10-08-eidolon-deploy-process.markdown

```diff
@@ -0,0 +1,192 @@
+
+```rb
+changelog_contents = File.read(changelog_filename)
+existing_releases = changelog_contents.split('releases:').last
+this_release = changelog_yaml['upcoming'].map{ |note| note.prepend '  ' }.join("\n")
+changelog_contents = <<-EOS
+upcoming:
+releases:
+- version: #{version}
+  date: #{Time.new.strftime("%B %d %Y")}
+  notes:
+#{this_release}
+#{existing_releases}
+EOS
+
+File.open(changelog_filename, 'w') { |file| file.puts changelog_contents }
+```
+
+At this point, we're ready to start the actual deploy process. First we need to download the provisioning profiles, which is only one step with Fastlane:
+
+```rb
+sigh
+```
+
+Next we build our app using `gym`. We need to use the legacy build API, I can't remember why.
+
+```rb
+gym(
+  scheme: "Kiosk",
+  export_method: 'enterprise',
+  use_legacy_build_api: true
+)
+```
+
+With our build finished, we upload to Hockey.
+
+```rb
+hockey(
+  api_token: hockey_api_token,
+  notes: release_notes
+)
+```
+
+Okay, our build is deployed. Time to let the team know there's a new version available:
+
+```rb
+slack(
+  message: "There is a new version of the Kiosk app available. Download it at http://artsy.net/kioskbeta",
+  success: true,        # optional, defaults to true
+  payload: {            # optional, lets you specify any number of your own Slack attachments
+    'Version' => version,
+    'What\'s new' => release_notes,
+  },
+  default_payloads: [],
+)
+```
+
+`default_payloads` needs to be empty I think, I can't remember why. Seems like "I can't remember why" is a common theme here...
+
+Before committing the changes we've made to the changelog and Info.plist files, we need to clean any build artefacts.
+
+```rb
+clean_build_artifacts
+```
+
+Finally, we commit, tag the build, and push to GitHub. Fastlane's built-in commands to commit to git reject any changes except to Info.plist files, and we've modified the changelog, so I used `sh` and used git directly.
+
+```rb
+sh "git add .. ; git commit -m 'Deploying version #{version}.'"
+add_git_tag tag: version
+push_to_git_remote
+```
+
+And that's it! With one terminal command, we've done all the following:
+
+- Verified version number format.
+- Verified the local environment is set up to deploy.
+- Verified API keys used by the app aren't empty.
+- Incremented the build number and version.
+- Updated the changelog.
+- Built and signed the app.
+- Uploaded the build to Hockey.
+- Posted a notification to Slack.
+- Tagged the release and pushed to GitHub.
+
+## Lessons Learned
+
+Automating Eidolon deploys was one of the first automated deploys we built on Artsy's iOS team. Now, based on Eidolon's successful deploy process, all our iOS deploys are automated.
```

Comment:
It _could_ be wroth talking about why we choose to do it with Eidolon ( difficulty in beta meant constant redeploys etc, as we needed to provide more betas that before, deploys on the night blah blah )